core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorCapabilities.java

```diff
@@ -16,4 +16,5 @@
 public enum ConnectorCapabilities
 {
     NOT_NULL_COLUMN_CONSTRAINT,
+    REPORT_BYTES_WRITTEN,
```

Comment:
This should probably be a method in `ConnectorMetadata`, instead. An enum doesn't allow connectors to customize the response based on other arguments (e.g., some connectors may be able to report bytes written for certain tables but not for others, such as connectors that interact with multiple backend systems).

See, for instance, `ConnectorMetadata.supportsMetadataDelete` and `ConnectorMetadata.supportsMissingColumnsOnInsert`.

Reply:
> some connectors may be able to report bytes written for certain tables but not for others, such as connectors that interact with multiple backend systems

@martint do you have examples?

It theory it can be even more complex. for example in hive connector writing to partitioned table, it *could* depend on the file format of particular partitions the query ends up writing to, so we cannot fully future-proof here.

I agree that per-table declaration is more flexible the per-connector.

BTW 
wouldn't same concern apply to `NOT_NULL_COLUMN_CONSTRAINT`?
if connector writes to different systems, some of them may not be able to retain "not null" column information.

Reply:
@martint , I can change the implemntation, obviously. It is actually flexible to decide to enable / disable scaleWriters basing on `TableHandle`. 

But, as far I know, the target table handle can be changed during optimizations - from what I know this why there is the `getWriterTarget` method in the `BeginTableWrite` class. As a result, the `BeginTableWrite` is an optimization that is executed as the last. And it is our constraint. 

The decision wheter to enable / disable `scaleWriter` we make in `AddExchanges`. If we want to base on target `TableHandle` we need to remember that target table could be changed by other optimization. So, it imply that AddExchanges must be nearly the last optmization. 

This is how I understand it. It is very possible that I misunderstand something.

Reply:
> wouldn't same concern apply to NOT_NULL_COLUMN_CONSTRAINT

Yes, but that's already there -- this PR is not about changing that (although we should, at some point)

> But, as far I know, the target table handle can be changed during optimizations - from what I know this why there is the getWriterTarget method in the BeginTableWrite class. As a result, the BeginTableWrite is an optimization that is executed as the last. And it is our constraint.

It can, but this is mostly to mark the beginning of the execution. That's currently implemented as an optimization rule because of lack of lifecycle hooks in the engine. It shouldn't be a problem for this purpose.

Also, having a method makes it easier to change the expected default behavior in the future. We can say, at some point, that connectors must report bytes written (and have a default method implementation that returns true), and otherwise, implement the method and have it return false. With an enum, it's not possible to do so. Connectors will have to return that value forever, or we'll have to introduce a DOES_NOT_REPORT_BYTES_WRITTEN to indicate the opposite.

Reply:
@martint , I agree that it is more flexible to implement it as method (not enum). 

To sum up, I can implement it as method `f` in `ConnectorMetadata` but I do not know if this method can take `tableHandle` reliably? Namely, if this method gets `X` as parameter it seems to be potentially errorneous - as we said - the table handle could change. Becasue the method decides for table `X`, but actual write could be made to table `Y`. This is what is a concern of mine.

Reply:
That shouldn't be a concern. The same could happen if the connector reports that capability (as an enum) but fails to provide the bytes written for the eventual table handle (e.g., if it's just a pass-through to another connector). To be safe, we could add a check to ensure the table handle reports bytes written somewhere before the query starts. That would help surface bugs in connectors.

In the future (when we have an exploratory optimizer, we could even model this as a trait that can be used to explore a plan that has scale-writers enabled)

Reply:
Ok - I can rewrite it.

> To be safe, we could add a check to ensure the table handle reports bytes written somewhere before the query starts. That would help surface bugs in connectors.

It would be great but we cannot do it easily without redesigning how does it work in Trino. Namely, the some `partitionScheme` is decided during optimization. Assume, that we detect that target table handle does not report written bytes after the optimization decided to use scaleWriters. But, what can we do then? We should choose another `partitionScheme` - this is exactly what optmizations do - it smells like strong redesing.

----------------------------------

This what I propose: 

To check whether the connector reports bytes written we need to have kind of annotation (like enum for `ConnectorCapabilities`). Is it ok for you that I implement a method:

```
boolean ConnectorMetadata.supportsScaleWriters(TableHandle targetTable) {
// base implementation checks whether there is a `REPORTS_BYTES_WRITTEN` in connector capabilities.  
}

```
If any connector has more "complex" policy of enabling scale writers, then the connector should override this method. In this way we have a flexible approach.

Alternatively, I can implement it as 


```
boolean ConnectorMetadata.supportsScaleWriters(TableHandle targetTable) {
// base implementation returns false. 
}
```

Reply:
> Assume, that we detect that target table handle does not report written bytes after the optimization decided to use scaleWriters.

We'd just fail the query. 

> If any connector has more "complex" policy of enabling scale writers, then the connector should override this method.

That's not necessary. First, connectors don't know anything about scaled writers. Second, as long as they report the number of bytes, writer scaling will just work -- no need for anything more complicated. Said another way, how would a connector writer decide whether scaled writers are supported?

Reply:
> we want to have a method to be more flexible.

More flexible in terms in whether the connector can report bytes written for a given table, but not necessarily making connectors aware of scaled writers as an engine feature they need to actively support.

Reply:
Yes, this how I understand it :)